extraConfigDpu.py

```diff
@@ -231,7 +231,7 @@ def ExtraConfigDpu(cc: ClustersConfig, cfg: ExtraConfigArgs, futures: dict[str,
     [f.result() for (_, f) in futures.items()]
     logger.info("Running post config step to start DPU operator on IPU")
 
-    repo = cfg.dpu_operator_path
+    repo = cfg.dpu_operator_path or "/root/dpu-operator"
```

Comment:
Do we really need this check?

We already set the default here: https://github.com/bn222/cluster-deployment-automation/blob/main/clustersConfig.py#L57

Reply:
The line you quote is

```
dpu_operator_path: Optional[str] = "/root/dpu-operator"
```

so mypy will think this value can be `None`.

The check is at least there to silence mypy warning.

And that `Optional[str]` is there for good reasons, because this parameter only makes sense with `ExtraConfigArgs` of `name: dpu_operator_{dpu,host}`.

What is ugly in this patch is that the default path is referenced twice. That will soon change on upcoming `th/clustersConfig`, where ClustersConfig ensures that the value is set if-and-only-if the `name` of the `ExtraConfigArgs` is `in ("dpu_operator_dpu", "dpu_operator_host")` ([here](https://github.com/thom311/cluster-deployment-automation/blob/5a7b6ad0180a8e898b3f7382136dfc8c06f2adda/clustersConfig.py#L255)). At that point, users can trust the validation and just do [unwrap()](https://github.com/thom311/cluster-deployment-automation/blob/5a7b6ad0180a8e898b3f7382136dfc8c06f2adda/extraConfigDpu.py#L196).


Point being, this will soon improve (to extend the validation/normalization in `clustersConfig.py` and then use `unwrap()`).

Reply:
But why is dpu_operator_path of type `Optional[str]` instead of `str`? If we implement defaults differently, then we can have it be a `str` instead and mypy will also be happy.

Reply:
As it is at the moment, it just as well could be a `str`, since the value is (almost) always set to a string.

However, with `ExtraConfigArgs` we have one type for all the various pre/post config types, and most fields only make sense for the right "name". For example,

```yaml
     postconfig:
       - name: dpu_operator_dpu
         dpu_operator_path: /some/where
```
make sense, but
```yaml
     postconfig:
       - name: microshift
         dpu_operator_path: /some/where
```
makes no sense and should be rejected as invalid configuration.

Furthermore, an `ExtraConfigArgs` with `name=="microshift"` should have the `dpu_operator_path` set to `None` to represent that this value is not there (and makes no sense). Having the property a `Optional[str]` is quite right.

Reply:
have a string -> string opaque map has it's pros too: a few lines of code, easy to add things. You lose safety but you gain developer speed. post configs are a hack anyway. I would keep things simple.

Reply:
Why are post-configs a hack? How would you provide those features otherwise? These seems useful features.

---

Using a opaque strdict map goes into the opposite direction of what was done in 9714ec0a6b631beac69737de33fd4f71db3f932b and what will further expand with https://github.com/bn222/cluster-deployment-automation/pull/254.

It's rather useful to have the linter help you. It's also useful to get feedback about invalid user configuration. Especially, since there is no documentation otherwise (the parsing/validation code sufficiently acts as documentation). 

If there was no validation, wouldn't then the caller still need to check the value (and fallback to a default)? If the caller still does such validation, it doesn't actually safe anything. It still is done somewhere, just at a worse place. And if callers don't do any validation, that sounds unmaintainable.

On #254, the handling of the typed `dpu_operator_path` property costs 14 lines of code (not counting blank lines). And two lines that use it via `unwrap(cfg.dpu_operator_path)`. If a user misconfigures something or scratches their head how to configure something, they loose more time than what is gained by developer speed.

Reply:
Because CDA is "cluster deployment automation", not "automation to deploy things after a cluster has been deployed". The latter should be part of the project that uses CDA.

Reply:
what does "keep [it] opaque" mean? It is *already* not opaque.

This is a tiny patch that fixes an actual bug. `dpu_operator_path` is a typed `Optional[str]`, so it needs something to shut up the linter, the "or PATH" is that something.

The branch does not do the large task to convert `ExtraConfigArgs` back to an opaque str->str dict, which also would be the wrong direction and go against https://github.com/bn222/cluster-deployment-automation/pull/254 .

Reply:
There is currently no agreement yet on the state of https://github.com/bn222/cluster-deployment-automation/pull/254

I looked into that PR and I can see:

```
-    dpu_operator_path: Optional[str] = "/root/dpu-operator"
+    dpu_operator_path: Optional[str]
```

in `ExtraConfigArgs`.

Why do we remove the default there? I'm wondering if we can make it a simple string and set the default:

`dpu_operator_path: str = "/root/dpu-operator"`

Reply:
> Why is the PR proposing to remove the default there?

The dataclass should not have default values in that scheme. But there is still a default value determined during `ExtraConfigArgs.parse()`  [here](https://github.com/thom311/cluster-deployment-automation/blob/8fc817803dd031e7e94a9b95dea5a496da0d2ad2/clustersConfig.py#L265). In particular, the default depends on the `name` of the extra config. For `name: dpu_operator_{host,dpu}` it's `"/root/dpu-operator"`, otherwise it's `None`.

> I'm wondering if we can make it a simple string

We can, but that is wrong for reasons said earlier. The `ExtraConfigArgs` represent all types of preconfig/postconfig, but the option `dpu_operator_path` only makes sense with `dpu_operator_{host,dpu}`. Where it doesn't make sense, the value is enforced to be `None`, while it makes sense, it's always set. The code that uses the parameter (in `extraConfigDPU.py`), relies on that validation and merely uses `unwrap()` to get the value ([here](https://github.com/thom311/cluster-deployment-automation/blob/8fc817803dd031e7e94a9b95dea5a496da0d2ad2/extraConfigDpu.py#L156)).

Reply:
right, let's simplify the types and have it not be `Optional[str]` just for the sake that it doesn't always make sense. We're going to keep `str` for all the types and use defaults there instead of passing around optionals and having to `unwrap` it. Please also change https://github.com/bn222/cluster-deployment-automation/pull/254 in addition to this PR.

Reply:
the optional is now gone with this patch because it doesn't matter here.

(note that a follow-up patch will restore it, because it does matter there. but that is a separate topic).

This is thus handled (for now and as far as this PR is concerned).